core/action/service.go

```diff
@@ -19,11 +44,28 @@ func (s Service) Get(ctx context.Context, id string) (Action, error) {
 }
 
 func (s Service) Create(ctx context.Context, action Action) (Action, error) {
+	currentUser, err := s.userService.FetchCurrentUser(ctx)
+	if err != nil {
+		return Action{}, fmt.Errorf("%w: %s", user.ErrInvalidEmail, err.Error())
+	}
+
 	newAction, err := s.repository.Create(ctx, action)
 	if err != nil {
 		return Action{}, err
 	}
 
+	go func() {
+		ctx := pkgctx.WithoutCancel(ctx)
+		actionLogData := newAction.ToActionLogData()
+		var logDataMap map[string]interface{}
+		if err := mapstructure.Decode(actionLogData, &logDataMap); err != nil {
+			s.logger.Errorf("%s: %s", ErrLogActivity.Error(), err.Error())
+		}
+		if err := s.activityService.Log(ctx, auditKeyActionCreate, currentUser.ID, logDataMap); err != nil {
+			s.logger.Errorf("%s: %s", ErrLogActivity.Error(), err.Error())
+		}
```

Comment:
We can pass the log data struct directly to the `Log` function and the `Log` function 4th argument accept any and do the mapstructure Decode inside the Log function.
Therefore, the decoding would only be done in a single place (inside the Log function only).